knimin/lib/data_access.py

```diff
@@ -3165,96 +3031,130 @@ def read_sample_plate(self, plate_id):
             TRN.add(sql, [plate_id])
             return dict(TRN.execute_fetchindex()[0])
 
-    def _sample_plate_layout_exists(self, id):
-        """Checks whether the layout of a sample plate exists
-
-        The layout of a sample plate exists when at least one sample-to-well
-        record exists.
-
-        Parameters
-        ----------
-        id : int
-            ID of the sample plate whose layout is to be checked
-
-        Returns
-        ------
-        bool
-            Whether the layout exists
-        """
-        with TRN:
-            sql = """SELECT EXISTS (SELECT 1 FROM pm.sample_plate_layout
-                                    WHERE sample_plate_id = %s)"""
-            TRN.add(sql, [id])
-            return TRN.execute_fetchlast()
-
-    def _clear_sample_plate_layout(self, id):
+    def _clear_sample_plate_layout(self, sample_plate_id):
         """Deletes the entire layout of a sample plate
 
         Parameters
         ----------
-        id : int
+        sample_plate_id : int
             ID of the sample plate whose layout is to be deleted
         """
         with TRN:
             sql = """DELETE FROM pm.sample_plate_layout
                      WHERE sample_plate_id = %s"""
-            TRN.add(sql, [id])
+            TRN.add(sql, [sample_plate_id])
             TRN.execute()
 
-    def write_sample_plate_layout(self, id, layout):
+    def write_sample_plate_layout(self, sample_plate_id, layout):
         """Writes the layout of a sample plate
 
         Parameters
         ----------
-        id : int
+        sample_plate_id : int
             ID of the sample plate whose layout is to be written
-        layout : list of dict
-            {sample_id : str, col : int, row : int, name : str, notes : str}
-            A list of sample-to-well records, each of which includes:
-            Sample ID, column number, row number, name, and notes
+        layout : list of list of dict
+            A 2-D matrix storing the per well information in a dict with the
+            format: {'sample_id': str, 'name': str, 'notes': str}
+
+        Raises
+        ------
+        ValueError
+            If the sample plate doesn't exist
+            If the given layout doesn't construct a valid layout
+            If the given layout doesn't match the sample plate type
         """
         with TRN:
-            self._sample_plate_exists(id)
-            if self._sample_plate_layout_exists(id):
-                self._clear_sample_plate_layout(id)
-            sql = """INSERT INTO pm.sample_plate_layout (sample_plate_id,
-                        sample_id, col, row, name, notes)
+            self._sample_plate_exists(sample_plate_id)
+
+            # Check if the given layout forms a valid layout
+            l_rows = len(layout)
+            l_cols = len(layout[0])
+            # Check that all rows have the same number of columns
+            # for r in layout
+            if any([len(r) != l_cols for r in layout]):
```

Comment:
Correct, wells need to be defined, but just need to pass None values to identify an empty well